atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/invariant/DurableWritesInvariantTest.java

```diff
@@ -0,0 +1,123 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.invariant;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.palantir.atlasdb.factory.TransactionManagers;
+import com.palantir.atlasdb.keyvalue.api.Namespace;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.ConflictHandler;
+import com.palantir.atlasdb.transaction.api.TransactionManager;
+import com.palantir.atlasdb.workload.store.AtlasDbTransactionStore;
+import com.palantir.atlasdb.workload.store.ImmutableWorkloadCell;
+import com.palantir.atlasdb.workload.store.TableWorkloadCell;
+import com.palantir.atlasdb.workload.store.WorkloadCell;
+import com.palantir.atlasdb.workload.transaction.WriteTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.ImmutableWitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedDeleteTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedWriteTransactionAction;
+import com.palantir.atlasdb.workload.util.AtlasDbUtils;
+import com.palantir.atlasdb.workload.workflow.ImmutableWorkflowHistory;
+import com.palantir.atlasdb.workload.workflow.WorkflowHistory;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class DurableWritesInvariantTest {
+
+    private static final String TABLE = "foo";
+    private static final TableReference TABLE_REFERENCE = TableReference.create(Namespace.DEFAULT_NAMESPACE, TABLE);
+
+    private static final WorkloadCell WORKLOAD_CELL_1 = ImmutableWorkloadCell.of(1, 2);
+    private static final WorkloadCell WORKLOAD_CELL_2 = ImmutableWorkloadCell.of(3, 4);
+    private static final Integer VALUE_ONE = 5;
+    private static final Integer VALUE_TWO = 3;
+
+    private TransactionManager manager;
+    private AtlasDbTransactionStore store;
+
+    @Before
+    public void before() {
+        manager = TransactionManagers.createInMemory(Set.of());
+        store = AtlasDbTransactionStore.create(
+                manager, Map.of(TABLE_REFERENCE, AtlasDbUtils.tableMetadata(ConflictHandler.SERIALIZABLE)));
+    }
+
+    @Test
+    public void undeletedCellsAreFoundAndDeletedAreIgnored() {
+        AtomicReference<Map<TableWorkloadCell, Integer>> undeletedCells = new AtomicReference<>();
+
+        DurableWritesInvariant durableWritesInvariant = new DurableWritesInvariant(_ignore -> {}, undeletedCells::set);
+        List<WitnessedTransaction> witnessedTransactions = List.of(
+                ImmutableWitnessedTransaction.builder()
+                        .addActions(WitnessedDeleteTransactionAction.of(TABLE, WORKLOAD_CELL_1))
```

Comment:
Writing a witnessed transaction is super painful, but I think we'll have to accept this for the time being :( 

Later on, we should make something that can use the builder pattern, and would greatly reduce verbosity here.

Such as:
```
WitnessTransactionHelper.newTransaction().delete(...).write(...).read(...).delete(...).multi(delete(...), read(...)).build()
```
which would automatically create start/commit timestamps for each action.